### ISSUE TITLE
Allow recursive types

### ISSUE THREAD

**Issue by DylanSp** [paarthenon/variant#5]
I ran into this issue when trying to create types for a lambda calculus interpreter. Repro example:

```
import variant, { variantList, VariantOf, fields, TypeNames } from "variant";

export const LambdaExpr = variantList([
  variant("Var", fields<{ ident: string }>()),
  variant("Apply", fields<{ func: LambdaExpr; arg: LambdaExpr }>()),
  variant("Lambda", fields<{ argName: string; body: LambdaExpr }>()),
]);

export type LambdaExpr<T extends TypeNames<typeof LambdaExpr> = undefined> = VariantOf<typeof LambdaExpr, T>;
```

The `const` declaration has this TypeScript error:

> 'LambdaExpr' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. ts(7022)

The type declaration has this TypeScript error:

> Type alias 'LambdaExpr' circularly references itself. ts(2456)

Is there a way to create recursive variant types that I'm missing? If not, would it be possible to add support for them to this library?

**Comment by paarthenon** [comment]
Hey!

Apologies for not responding for so long.

I haven't been able to make recursive variants work yet. It's something I'm experimenting with.

In the meantime, you can do things the "old" way by defining the type manually and then using variants to implement the functionality. It will still work with most of the functions in this library.

Here's an example for you!

```typescript
interface Var {
    ident: string;
}
interface Apply {
    func: LambdaExpr;
    arg: LambdaExpr;
}

interface Lambda {
    argName: string;
    body: LambdaExpr;
}

type LambdaExpr = 
    | {type: 'Var'} & Var 
    | {type: 'Apply'} & Apply
    | {type: 'Lamda'} & Lambda
;

export const LambdaExpr = variantList([
    variant('Var', fields<Var>()),
    variant('Apply', fields<Apply>()),
    variant('Lambda', fields<Lambda>()),
]);

const thing = LambdaExpr.Var({ident: 'foo'});
const lambda = LambdaExpr.Lambda({argName: 'bar', body: thing});

export const printExpression = (expr: LambdaExpr): string => match(expr, {
    Var: ({ident}) => `Var:${ident}`,
    Apply: ({arg, func}) => `Apply (${printExpression(arg)},  ${printExpression(func)})`,
    Lamda: ({argName, body}) => `Lambda:${argName} (${printExpression(body)})`,
});
```

**Comment by DylanSp** [comment]
That's unfortunate that recursive variants don't currently work; it pretty much rules out using this library for some of my projects. Let us know if you can get them working, it'd be great to see.

**Comment by paarthenon** [comment]
I understand completely. I'll @mention you if I manage them. Thank you for trying the library!

**Comment by paarthenon** [comment]
Hi @DylanSp ,

I think I got it. Here is a working code sample from variant `2.0.0-beta.2`.

```typescript
// type def
type LambdaExpr =
    | Variant<'Var', {ident: string}>
    | Variant<'Apply', {func: LambdaExpr, arg: LambdaExpr}>
    | Variant<'Lambda', {argName: string, body: LambdaExpr}>
;

// factory creation
const LambdaExpr = typedVariant<LambdaExpr>({
    Apply: pass,
    Lambda: pass,
    Var: pass,
});

// consumer function
function exprToString(expr: LambdaExpr): string {
    return match(expr, {
        Var: ({ident}) => `(Var: ${ident})`,
        Apply: ({func, arg}) => `(Apply [${exprToString(arg)}] to [${exprToString(func)}])`,
        Lambda: ({argName, body}) => `(Lambda (${argName}): [${body}])`,
    });
}

// creation and use
const expression = LambdaExpr.Lambda({argName: 'alpha', body: LambdaExpr.Var({ident: 'beta'})});
console.log(exprToString(expression));
```

I hope that meets what you're looking for? This change also includes support for **generic** variants.

https://github.com/paarthenon/variant/issues/2#issuecomment-708243260

See that comment for more details.

****

If you'd like to try out these (undocumented) changes yourself,

    npm i variant@beta

**Comment by paarthenon** [comment]
Hello again @DylanSp It's been a couple of weeks and variant 2.0.1 including the fix for this issue has been merged in. I'm going to be closing the issue for now. Thank you for bringing this up! Please feel free to reopen this issue or a new one if this approach doesn't address your needs.

**Comment by DylanSp** [comment]
Looks great, @paarthenon ! I'll have to give it a try when the opportunity presents itself, but it looks eminently usable.

**Comment by paarthenon** [comment]
I haven't updated the docs for this, so I wanted to post here in case it's relevant to you. These changes are available in `2.0.2-beta.6`

I added some better handling for recursive generics:

```typescript
type Tree<T> =
    | Variant<'Branch', {left: Tree<T>, right: Tree<T>}>
    | Variant<'Leaf', {payload: T}>
;

const Tree = genericVariant(({A}) => ({
    Branch: fields<{left: Tree<typeof A>, right: Tree<typeof A>}>(),
    Leaf: payload(A),
}));

const leaf = Tree.Leaf(4); // no need for separate numTree and animalTree constructors.
```
Attempting to use disparate types raises an error, as expected.

![image](https://user-images.githubusercontent.com/10353928/98486880-e782b200-21ed-11eb-8ea4-4c3deebbc84d.png)
If, however, the tree was created with `left` and `right` branches set to the generic terms `A` and `B` respectively, this no longer raises an error.

p.s. the `Variant<string, {}?>` type is very simply a stand-in for `{type: string} & {}?`. I encourage it for ergonomics, and because it's a possible entry point for future 'magic' for me, but it's not necessary.
